src/responder/nss/nss_protocol_sid.c

```diff
@@ -87,13 +87,115 @@ nss_get_id_type(struct nss_cmd_ctx *cmd_ctx,
+    enum sss_id_type ltype;
+
+    if (result->count == 1) {
+        *_sid = ldb_msg_find_attr_as_string(result->msgs[0],
+                                            SYSDB_SID_STR, NULL);
+        if (*_sid == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "Missing SID.\n");
+            return EINVAL;
+        }
+        return nss_get_id_type(cmd_ctx, result, _type);
+    }
+
+    for (c = 0; c < result->count; c++) {
+        ret = find_sss_id_type(result->msgs[c],
+                               false /* we are only interested in the type
+                                      * of the object, so mpg setting can
+                                      * be ignored */,
+                               &ltype);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "Unable to find ID type, ignored [%d][%s].\n",
+                  ret, sss_strerror(ret));
+            continue;
+        }
+
+        if (ltype == SSS_ID_TYPE_GID) {
+            tmp = ldb_msg_find_attr_as_string(result->msgs[c],
+                                               SYSDB_SID_STR, NULL);
+            if (tmp == NULL) {
+                continue;
+            }
+            if (tmp != NULL && group_sid != NULL) {
+                DEBUG(SSSDBG_OP_FAILURE,
+                      "Search for SID found multiple users with a SID, "
+                      "request failed.\n");
+                return EINVAL;
+            }
+            group_sid = tmp;
+            group_gid = ldb_msg_find_attr_as_uint64(result->msgs[c],
+                                                    SYSDB_GIDNUM, 0);
+        } else {
+            tmp = ldb_msg_find_attr_as_string(result->msgs[c],
+                                              SYSDB_SID_STR, NULL);
+            if (tmp == NULL) {
+                continue;
+            }
+            if (tmp != NULL && user_sid != NULL) {
+                DEBUG(SSSDBG_OP_FAILURE,
+                      "Search for SID found multiple users with a SID, "
+                      "request failed.\n");
+                return EINVAL;
+            }
+            user_sid = tmp;
+            user_uid = ldb_msg_find_attr_as_uint64(result->msgs[c],
+                                                   SYSDB_UIDNUM, 0);
+            user_gid = ldb_msg_find_attr_as_uint64(result->msgs[c],
+                                                   SYSDB_GIDNUM, 0);
+        }
+    }
+
+    if (user_sid == NULL && group_sid == NULL) {
+        /* No SID in the results */
+        return ENOENT;
+    } else if (user_sid != NULL && group_sid == NULL) {
+        /* There is only one user with a SID in the results */
+        *_sid = user_sid;
+        *_type = SSS_ID_TYPE_UID;
+    } else if (user_sid == NULL && group_sid != NULL) {
+        /* There is only one group with a SID in the results */
+        *_sid = group_sid;
+        *_type = SSS_ID_TYPE_GID;
+    } else if (user_sid != NULL && group_sid != NULL && user_uid != 0
```

Comment:
If this is enough to be sure that we are dealing with manually created user?

Reply:
Yes, as the comment says it is about a user with a manually created user private group and this is covered by the condition.